staging/src/k8s.io/kubectl/pkg/kuberc/kuberc.go

```diff
@@ -0,0 +1,408 @@
+// arguments based on user's kuberc configuration.
+type PreferencesHandler interface {
+	AddFlags(flags *pflag.FlagSet)
+	Apply(rootCmd *cobra.Command, args []string, errOut io.Writer) ([]string, error)
+}
+
+// Preferences stores the kuberc file coming either from environment variable
+// or file from set in flag or the default kuberc path.
+type Preferences struct {
+	getPreferencesFunc func(kuberc string) (*config.Preference, error)
+
+	aliases map[string]struct{}
+}
+
+// NewPreferences returns initialized Prefrences object.
+func NewPreferences() PreferencesHandler {
+	return &Preferences{
+		getPreferencesFunc: DefaultGetPreferences,
+		aliases:            make(map[string]struct{}),
+	}
+}
+
+type aliasing struct {
+	args    []string
+	flags   []config.CommandOverrideFlag
+	command *cobra.Command
+}
+
+// AddFlags adds kuberc related flags into the command.
+func (p *Preferences) AddFlags(flags *pflag.FlagSet) {
+	flags.String("kuberc", "", "Path to the kuberc file to use for preferences.")
+}
+
+// Apply firstly applies the aliases in the preferences file and secondly overrides
+// the default values of flags.
+func (p *Preferences) Apply(rootCmd *cobra.Command, args []string, errOut io.Writer) ([]string, error) {
+	if len(args) <= 1 {
+		return args, nil
+	}
+
+	kubercPath := getExplicitKuberc(args)
+	kuberc, err := p.getPreferencesFunc(kubercPath)
+	if err != nil {
+		return args, fmt.Errorf("kuberc error %w", err)
+	}
+
+	if kuberc == nil {
+		return args, nil
+	}
+
+	err = validate(kuberc)
+	if err != nil {
+		return args, err
+	}
+
+	args, err = p.applyAliases(rootCmd, kuberc, args, errOut)
+	if err != nil {
+		return args, err
+	}
+	err = p.applyOverrides(rootCmd, kuberc, args, errOut)
+	if err != nil {
+		return args, err
+	}
+	return args, nil
+}
+
+// applyOverrides finds the command and sets the defaulted flag values in kuberc.
+func (p *Preferences) applyOverrides(rootCmd *cobra.Command, kuberc *config.Preference, args []string, errOut io.Writer) error {
+	args = args[1:]
+	cmd, _, err := rootCmd.Find(args)
+	if err != nil {
+		fmt.Fprintf(errOut, "Warning: command not found %v\n", err)
+		return nil
+	}
+
+	for _, c := range kuberc.Overrides {
+		parsedCmds := strings.Fields(c.Command)
+		overrideCmd, _, err := rootCmd.Find(parsedCmds)
+		if err != nil {
+			fmt.Fprintf(errOut, "Warning: command %q not found to set kuberc override\n", c.Command)
+			continue
+		}
+		if overrideCmd.Name() != cmd.Name() {
+			continue
+		}
+
+		if _, ok := p.aliases[cmd.Name()]; ok {
+			return fmt.Errorf("alias %s can not be overridden", cmd.Name())
+		}
+
+		// This function triggers merging the persistent flags in the parent commands.
+		_ = cmd.InheritedFlags()
+
+		for _, fl := range c.Flags {
+			existingFlag := cmd.Flag(fl.Name)
+			if existingFlag == nil {
+				return fmt.Errorf("invalid flag %s for command %s", fl.Name, c.Command)
+			}
+			if searchInArgs(existingFlag.Name, existingFlag.Shorthand, cmd.Flags(), args) {
```

Comment:
rather than passing in `cmd.Flags()`, prefer accumulating the set of shortnames from the flags once and passing that in; that'll simplify `searchInArgs` for unit testing and avoid repeatedly accumulating as we loop through flags

Reply:
Updated.